homeassistant/components/isy994.py

```diff
@@ -44,190 +43,352 @@
+                        if domain != 'binary_sensor':
+                            actions = entity_folder[KEY_ACTIONS]
+                            assert actions.dtype == 'program', 'Not a program'
+                        else:
+                            actions = None
+                    except (AttributeError, KeyError, AssertionError):
+                        _LOGGER.warning("Program entity '%s' not loaded due "
+                                        "to invalid folder structure.",
+                                        entity_folder.name)
+                        continue
+
+                    entity = (entity_folder.name, status, actions)
+                    hass.data[ISY994_PROGRAMS][domain].append(entity)
+
+
+def _categorize_weather(hass: HomeAssistant, climate) -> None:
     """Categorize the ISY994 weather data."""
-    global WEATHER_NODES
-
-    climate_attrs = dir(ISY.climate)
-    WEATHER_NODES = [WeatherNode(getattr(ISY.climate, attr), attr,
-                                 getattr(ISY.climate, attr + '_units'))
+    climate_attrs = dir(climate)
+    _LOGGER.debug(climate)
+    _LOGGER.debug(climate_attrs)
+    weather_nodes = [WeatherNode(getattr(climate, attr),
+                                 attr.replace('_', ' '),
+                                 getattr(climate, '{}_units'.format(attr)))
                      for attr in climate_attrs
-                     if attr + '_units' in climate_attrs]
+                     if '{}_units'.format(attr) in climate_attrs]
+    hass.data[ISY994_WEATHER].extend(weather_nodes)
 
 
 def setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up the ISY 994 platform."""
+    if ISY994_NODES not in hass.data:
+        hass.data[ISY994_NODES] = {}
+    for domain in SUPPORTED_DOMAINS:
+        if domain not in hass.data[ISY994_NODES]:
+            hass.data[ISY994_NODES][domain] = []
+
+    if ISY994_WEATHER not in hass.data:
+        hass.data[ISY994_WEATHER] = []
+
+    if ISY994_PROGRAMS not in hass.data:
+        hass.data[ISY994_PROGRAMS] = {}
+        for domain in SUPPORTED_DOMAINS:
+            if domain not in hass.data[ISY994_PROGRAMS]:
+                hass.data[ISY994_PROGRAMS][domain] = []
+
     isy_config = config.get(DOMAIN)
 
     user = isy_config.get(CONF_USERNAME)
     password = isy_config.get(CONF_PASSWORD)
     tls_version = isy_config.get(CONF_TLS_VER)
     host = urlparse(isy_config.get(CONF_HOST))
-    port = host.port
-    addr = host.geturl()
-    hidden_identifier = isy_config.get(
-        CONF_HIDDEN_STRING, DEFAULT_HIDDEN_STRING)
-    sensor_identifier = isy_config.get(
-        CONF_SENSOR_STRING, DEFAULT_SENSOR_STRING)
-
-    global HIDDEN_STRING
-    HIDDEN_STRING = hidden_identifier
+    ignore_identifier = isy_config.get(CONF_IGNORE_STRING)
+    sensor_identifier = isy_config.get(CONF_SENSOR_STRING)
+    enable_climate = isy_config.get(CONF_ENABLE_CLIMATE)
 
     if host.scheme == 'http':
```

Comment:
These checks should be moved to a validator in the config schema.

Reply:
You could parse and insert the port in the config in the validator too.

Reply:
Maybe you can use `vol.Url()` as the first validator?

Reply:
Ah, I haven't actually looked in to the config schema framework. If it can handle the URL stuff that would be much better. I'll do some digging.

Reply:
@MartinHjelmare Do you have an example of what you're looking for? I looked in to this, and the config option is already being passes through `cv.url` which in turn uses `vol.Url()`. This code here isn't really doing validation, but just munging the URL in to the format that the PyISY library expects (all of this could/should be done inside PyISY if it just accepted a URL rather than a host, port, and https boolean, but it doesn't.)

I could remove the final else here, because cv.url already verified that we're either http or https, but I'm not sure anything more than that makes sense.

Reply:
```py
def validate_host(config):
    host = urlparse(config[CONF_HOST])
    if host.scheme == 'http':
        https = False
        port = host.port or 80
    elif host.scheme == 'https':
        https = True
        port = host.port or 443
    config[CONF_HOST] = host
    config[CONF_PORT] = port
    config[CONF_HTTPS] = https
    return config
```

Reply:
Ah, so this would be just moving it to a method that gets called from within `setup`, and injects in to the config Dict? It doesn't seem like this is something that happens inside the schema definition itself, right?

My concern with this is that it injects new config values in to the dict that didn't go through Voluptuous's Schema checks. A casual glance at the code (due to the new CONF consts) would also make it look like PORT and HTTPS can be provided as config options, even though they can't.

Reply:
The validator would be used in the config schema.
```py
CONFIG_SCHEMA = vol.Schema({
    DOMAIN: vol.Schema(vol.All({...}, validate_host)})
```